File: PE-common/src/main/java/nl/adaptivity/util/xml/NamespaceAddingStreamReader.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.util.xml;

import net.devrieze.util.StringUtil;
import nl.adaptivity.xml.XmlException;
import nl.adaptivity.xml.XmlReader;
import nl.adaptivity.xml.XmlStreaming.EventType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.xml.namespace.NamespaceContext;


/**
 * A streamreader that adds a namespace context as source for looking up namespace information for prefixes present
 * in the original, but do not have attached namespace information.
 *
 * Created by pdvrieze on 31/10/15.
 */
public class NamespaceAddingStreamReader extends XmlDelegatingReader {

  private final NamespaceContext mLookupSource;

  public NamespaceAddingStreamReader(final NamespaceContext lookupSource, final XmlReader source) {
    super(source);
    mLookupSource = lookupSource;
  }

  @Override
  public void require(final EventType type, @Nullable final CharSequence namespaceURI, @Nullable final CharSequence localName) throws
          XmlException {
    if (type != getEventType() ||
            (namespaceURI != null && !namespaceURI.equals(getNamespaceUri())) ||
            (localName != null && !localName.equals(getLocalName()))) {
      mDelegate.require(type, namespaceURI, localName);
    } {
      throw new XmlException("Require failed");
    }
  }

  @Override
  public String getNamespaceUri(final CharSequence prefix) throws XmlException {
    final String namespaceURI = mDelegate.getNamespaceUri(prefix);
    return namespaceURI != null ? namespaceURI : mLookupSource.getNamespaceURI(prefix.toString());
  }

  @Nullable
  @Override
  public CharSequence getAttributeValue(@Nullable final CharSequence namespaceURI, @NotNull final CharSequence localName) throws
          XmlException {

    for(int i=getAttributeCount()-1; i>=0; --i) {
      if ((namespaceURI==null || namespaceURI.equals(getAttributeNamespace(i))) && localName.equals(getAttributeLocalName(i))) {
        return getAttributeValue(i);
      }
    }
    return null;
  }

  @Override
  public CharSequence getAttributeNamespace(final int index) throws XmlException {
    final CharSequence attributeNamespace = mDelegate.getAttributeNamespace(index);
    return attributeNamespace!=null ? attributeNamespace : mLookupSource.getNamespaceURI(StringUtil.toString(mDelegate.getAttributePrefix(index)));
  }

  @Override
  public CharSequence getNamespaceUri() throws XmlException {
    final CharSequence namespaceURI = mDelegate.getNamespaceUri();
    return namespaceURI !=null ? namespaceURI : mLookupSource.getNamespaceURI(StringUtil.toString(mDelegate.getPrefix()));
  }

  public NamespaceContext getNamespaceContext() throws XmlException {
    return new CombiningNamespaceContext(mDelegate.getNamespaceContext(), mLookupSource);
  }
}

File: PE-common/common/src/main/kotlin/nl/adaptivity/util/xml/NamespaceAddingStreamReader.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.xmlutil.util

import nl.adaptivity.xmlutil.EventType
import nl.adaptivity.xmlutil.NamespaceContext
import nl.adaptivity.xmlutil.XmlException
import nl.adaptivity.xmlutil.XmlReader


/**
 * A streamreader that adds a namespace context as source for looking up namespace information for prefixes present
 * in the original, but do not have attached namespace information.
 *
 * Created by pdvrieze on 31/10/15.
 */
class NamespaceAddingStreamReader(private val lookupSource: NamespaceContext, source: XmlReader) : XmlDelegatingReader(
        source) {

    override val namespaceURI: String
        get() {
            val namespaceURI = delegate.namespaceURI
            return if(namespaceURI.isNotEmpty()) namespaceURI else lookupSource.getNamespaceURI(delegate.prefix) ?: ""
        }

    override val namespaceContext: NamespaceContext
        get() = CombiningNamespaceContext(delegate.namespaceContext, lookupSource)

    override fun require(type: EventType, namespace: String?, name: String?) {
        if (type !== eventType ||
            namespace != null && namespace != this.namespaceURI ||
            name != null && name != name) {
            delegate.require(type, namespace, name)
        }
        run { throw XmlException("Require failed") }
    }

    override fun getNamespaceURI(prefix: String): String? {
        val namespaceURI = delegate.getNamespaceURI(prefix)
        return namespaceURI ?: lookupSource.getNamespaceURI(prefix)
    }

    override fun getAttributeValue(nsUri: String?, localName: String): String? {

        for (i in attributeCount - 1 downTo 0) {
            if ((nsUri == null || nsUri == getAttributeNamespace(
                            i)) && localName == getAttributeLocalName(i)) {
                return getAttributeValue(i)
            }
        }
        return null
    }

    override fun getAttributeNamespace(index: Int): String {
        val attributeNamespace = delegate.getAttributeNamespace(index)
        return if(attributeNamespace.isNotEmpty()) attributeNamespace else lookupSource.getNamespaceURI(
            delegate.getAttributePrefix(index)) ?: ""
    }
}
